internal/charmstore/resources.go

```diff
@@ -0,0 +1,256 @@
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (s *Store) insertResource(doc *mongodoc.Resource) error {
+	if err := doc.Validate(); err != nil {
+		return errgo.Mask(err)
+	}
+	err := s.DB.Resources().Insert(doc)
+	if err != nil {
+		if mgo.IsDup(err) {
+			return errgo.WithCausef(nil, params.ErrDuplicateUpload, "")
+		}
+		return errgo.Notef(err, "cannot insert resource")
+	}
+	return nil
+}
+
+func (s *Store) storeResource(blob ResourceBlob) (string, error) {
+	name := bson.NewObjectId().Hex()
+	// TODO(ericsnow) We will finish this in a follow-up patch.
+	return name, nil
+}
+
+func (s *Store) setResource(entity *mongodoc.Entity, channel params.Channel, name string, revision int) error {
+	if channel == params.NoChannel {
+		return errgo.New("missing channel")
+	}
+	if channel == params.UnpublishedChannel {
+		return errgo.New("cannot publish to unpublished channel")
+	}
+
+	doc, err := s.resource(entity.BaseURL, name, revision)
+	if err != nil {
+		return errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+
+	if !charmHasResource(entity.CharmMeta, doc.Name) {
+		return errgo.Mask(err)
+	}
+
+	update := bson.D{{fmt.Sprintf("channelresources.%s.%s", channel, name), revision}}
+	if err := s.updateBaseEntity(entity.BaseURL, bson.D{{"$set", update}}); err != nil {
+		return errgo.Mask(err)
+	}
+	return nil
+}
+
+func (s *Store) latestResource(baseEntity *mongodoc.BaseEntity, channel params.Channel, name string) (*mongodoc.Resource, error) {
+	revision, err := s.latestResourceRevision(baseEntity, channel, name)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+	doc, err := s.resource(baseEntity.URL, name, revision)
+	if err != nil {
+		return nil, errgo.Mask(err, errgo.Is(resourceNotFound))
+	}
+	return doc, nil
+}
+
+func (s *Store) latestResourceRevision(baseEntity *mongodoc.BaseEntity, channel params.Channel, name string) (int, error) {
+	if channel == params.UnpublishedChannel {
+		var res mongodoc.Resource
+		err := s.DB.Resources().Find(bson.D{{"baseurl", baseEntity.URL}, {"name", name}}).Sort("-revision").One(&res)
+		if err == mgo.ErrNotFound {
+			return -1, resourceNotFound
+		}
+		if err != nil {
+			return -1, errgo.Mask(err)
+		}
+		return res.Revision, nil
+	}
+	revision, ok := baseEntity.ChannelResources[channel][name]
+	if !ok {
+		return -1, resourceNotFound
+	}
+	return revision, nil
+}
+
+func (s *Store) resource(url *charm.URL, name string, revision int) (*mongodoc.Resource, error) {
+	query := newResourceQuery(url, name, revision)
+	var doc mongodoc.Resource
+	err := s.DB.Resources().Find(query).One(&doc)
+	if err == mgo.ErrNotFound {
+		err = resourceNotFound
+		return nil, errgo.WithCausef(err, err, "")
+	}
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	if err := doc.Validate(); err != nil {
+		return nil, errgo.Notef(err, "got bad data from DB")
+	}
+	return &doc, nil
+}
+
+func (s *Store) publishedResources(curl *charm.URL, channel params.Channel) ([]*mongodoc.Resource, error) {
```

Comment:
I think I'd leave this out until we actually need it. As it is, it's neither tested nor used AFAICS.